e2e/exhaustive_suite_test.go

```diff
@@ -0,0 +1,80 @@
+// SPDX-FileCopyrightText: The RamenDR authors
+// SPDX-License-Identifier: Apache-2.0
+
+package e2e_test
+
+import (
+	"testing"
+
+	"github.com/ramendr/ramen/e2e/deployers"
+	"github.com/ramendr/ramen/e2e/testcontext"
+	"github.com/ramendr/ramen/e2e/workloads"
+)
+
+// Deployers = {"Subscription", "AppSet", "Imperative"}
+// Workloads = {"Deployment", "STS", "DaemonSet"}
+// Classes   = {"rbd", "cephfs"}
+
+func Exhaustive(t *testing.T) {
+	t.Helper()
+	t.Parallel()
+
+	deployment := &workloads.Deployment{}
+	deployment.Init()
+
+	Workloads := []workloads.Workload{deployment}
+
+	subscrition := &deployers.Subscription{}
+	subscrition.Init()
+
+	appset := &deployers.ApplicationSet{}
+	appset.Init()
+
+	Deployers := []deployers.Deployer{subscrition, appset}
+
+	for _, workload := range Workloads {
+		for _, deployer := range Deployers {
+			// assign workload and deployer to a local variable to avoid parallel test issue
+			// see https://go.dev/wiki/CommonMistakes
+			w := workload
+			d := deployer
+
+			t.Run(w.GetID(), func(t *testing.T) {
+				t.Parallel()
+				t.Run(d.GetID(), func(t *testing.T) {
+					t.Parallel()
+					testcontext.AddTestContext(t.Name(), w, d)
+					runTestFlow(t)
+				})
+			})
+		}
+	}
+}
+
+func runTestFlow(t *testing.T) {
+	t.Helper()
+
+	if !t.Run("Deploy", DeployAction) {
+		t.Fatal("Deploy failed")
```

Comment:
t.Fatal or something else on failures? Other tests can continue, this test has to be stopped. If there is no other construct to allow this test to fail and others to pass then we can stick with Fatal.

Reply:
for sub tests that belong to same test combination (workload * deployer), I would assume each sub test should not fail.  If one fails, then all subsequent tests would fail or entering a unknown/unsupported situation. 

For example:
if deploy fails, then we could not run failover, relocate since no app is deployed.
if failover fails, it may stuck into a state that no further failover or relocate could run.

So I think if one sub test fails, the subsequent tests in that test combination do not need run.

subtests belong to different test combinations (workload * deployer) should run in parallel without impacting each other. Failing one subtest in combination A does not impact subtest in combination B, unless it causes some infra failure that may impact all others, like it makes submariner not functioning.

Reply:
btw, `t.Fail` marks this function failed but continues execution for subsequent tests.

Reply:
Ack! As long as subtest A Fatal will not cause all subtests to stop/fail, we are good.